presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java

```diff
@@ -62,204 +53,75 @@ public UnnestBlockBuilder(Type type)
     public void resetInputBlock(Block block)
     {
         this.source = requireNonNull(block, "block is null");
-        // output and null-check have to be cleared because the source has changed
-        clearCurrentOutput();
-        nullFinder.resetCheck();
-    }
-
-    /**
-     * Resets all data structures used for producing current output block, except for the source block.
-     */
-    @VisibleForTesting
-    void clearCurrentOutput()
-    {
-        // clear dictionary block state
-        ids = new int[0];
-        positionCount = 0;
-        usingCopiedBlock = false;
-
-        // clear copied block state
-        outputBlockBuilder = null;
-        pageBuilderStatus = null;
+        nullFinder.resetCheck(block);
+        sourcePosition = 0;
     }
 
-    /**
-     * Prepares for a new output block.
-     * The caller has to ensure that the source is not null when this method is invoked.
-     */
-    public void startNewOutput(PageBuilderStatus pageBuilderStatus, int expectedEntries)
+    public Block buildOutputBlockWithoutNulls(int maxTotalEntriesForBatch)
     {
-        checkState(source != null, "source is null");
-        this.pageBuilderStatus = pageBuilderStatus;
-
-        // Prepare for a new dictionary block
-        this.ids = new int[expectedEntries];
-        this.positionCount = 0;
-        this.usingCopiedBlock = false;
-    }
+        int[] ids = new int[maxTotalEntriesForBatch];
 
-    public Block buildOutputAndFlush()
-    {
-        Block outputBlock;
-        if (usingCopiedBlock) {
-            outputBlock = outputBlockBuilder.build();
+        for (int j = 0; j < maxTotalEntriesForBatch; j++) {
+            ids[j] = sourcePosition++;
         }
-        else {
-            outputBlock = new DictionaryBlock(positionCount, source, ids);
-        }
-
-        // Flush stored state, so that ids can not be modified after the dictionary has been constructed
-        clearCurrentOutput();
 
-        return outputBlock;
+        return new DictionaryBlock(maxTotalEntriesForBatch, source, ids);
```

Comment:
Why is DictionaryBlock useful here? Why not just return a view into source block?

```
        Block output = source.getRegion(sourcePosition, batchSize);
        sourcePosition += batchSize;
        return output;
```